doc/src/dev/coding_style.rst

```diff
@@ -0,0 +1,217 @@
+      };
+
+- **Variable and function names** use `snake_case`.
+
+  .. code-block:: cpp
+
+      int net_count;
+      void estimate_wirelength();
+
+- **Private member variables and private methods** should end with an underscore `_`.
+
+  .. code-block:: cpp
+
+      class Example {
+      private:
+          int count_;
+          void update_state_();
+      };
+
+- **Enums** should use `enum class` instead of traditional enums. Enum names should start with `e_`.
+
+  .. code-block:: cpp
+
+      enum class e_heap_type {
+          INVALID_HEAP = 0,
+          BINARY_HEAP,
+          FOUR_ARY_HEAP,
+      };
+
+- **Trivial structs** that primarily store data without behavior should be prefixed with `t_`.
+
+  .. code-block:: cpp
+
+      struct t_dijkstra_data {
+          vtr::vector<RRNodeId, bool> node_expanded;
+          vtr::vector<RRNodeId, float> node_visited_costs;
+          std::priority_queue<PQ_Entry> pq;
+      };
+
+- **Source and header file names** use `snake_case` and should be short but descriptive.
+
+  Example: `router_lookahead_map.cpp`, `netlist_walker.h`
+
+
+Use of `auto`
+~~~~~~~~~~~~~
+
+Use `auto` only when the type is long, complex, or hard to write explicitly.
+
+Examples where `auto` is appropriate:
+
+.. code-block:: cpp
+
+    auto it = container.begin();  // Iterator type is long and not helpful to spell out
+
+    // The return type is std::vector<std::vector<std::pair<int, float>>>
+    auto matrix = generate_adjacency_matrix();
+
+    // Lambdas have unreadable and compiler-generated types — use auto for them
+    auto add = [](int x, int y) { return x + y; };
+
+
+Avoid `auto` when the type is simple and clear:
+
+.. code-block:: cpp
+
+    // Use type names when they are short and readable.
+    for (RRNodeId node_id : device_ctx.rr_graph.nodes()) {
+        t_rr_node_route_inf& node_inf = route_ctx.rr_node_route_inf[rr_id];
+    }
+
+    int count = 0;
+    std::string name = "example";
+    std::vector<int> numbers = {1, 2, 3};
+
+Avoid:
+
+.. code-block:: cpp
+
+    auto count = 0;                // Simple and obvious type
+    auto name = std::string("x");  // Hides a short, clear type
+
+    for (auto node_id : device_ctx.rr_graph.nodes()) {
+        // node_id is RRNodeId. Write it out for clarity.
+        auto& node_inf = route_ctx.rr_node_route_inf[rr_id];
+        // node_inf is t_rr_node_route_inf. Use the explicit type since it's simple and meaningful.
+    }
+
+Rationale: clear, explicit types help with readability and understanding. Avoid hiding simple types behind `auto`.
+
+
+
+Commenting Style
+~~~~~~~~~~~~~~~~
+
+Comments help readers understand the purpose, structure, and reasoning behind the code.
+This section outlines when and how to use comments in a consistent and helpful way.
+
+General Rules
+-------------
+
+- Focus on explaining *why* the code exists or behaves in a certain way.
+- Do not explain *what* the code is doing if it's already clear from the code itself.
```

Comment:
.h file should say what the code does; if not obvious explain why the code behaves that way as well.
.cpp file should explain how the code works if it is not obvious.